### PR TITLE
Support `COMMENT ON` partitioned column in HMS

### PR DESCRIPTION
**Pull request by ssheikin**
## Description

COMMENT ON partitioned column in File HMS and Thrift HMS was ignored. Whereas it works for non-partitioned columns.
Also it's already addressed for Glue HMS.
Adding support for  `COMMENT ON` partitioned column in File HMS and Thrift HMS

## Release notes

(x) Release notes are required, with the following suggested text:

```markdown
# Hive
* Add support for commenting on partitioned columns in Thrift Hive metastore. ({issue}`20264`)
```

### CODE REVIEW

**Comment by ssheikin** [comment]
cc @SemionPar

**Comment by findinpath** [comment]
build :red_circle: 

checkstyle
```
Error:  src/main/java/io/trino/plugin/hive/metastore/file/FileHiveMetastore.java:[788,5] (whitespace) EmptyLineSeparator: 'METHOD_DEF' should be separated from previous line.
```


please also add a meaningful description for the PR stating its purpose.

> Release notes are required. Please propose a release note for me.

Let's change this to choosing

> Release notes are required, with the following suggested text:

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/file/FileHiveMetastore.java [hunk 1]
```diff
@@ -736,26 +737,54 @@ public synchronized void commentTable(String databaseName, String tableName, Opt
     @Override
     public synchronized void commentColumn(String databaseName, String tableName, String columnName, Optional<String> comment)
     {
```
Comment (findinpath) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/file/FileHiveMetastore.java:739]: "Fix COMMENT ON partitioned column for FileHiveMetastore"

The commit message states that something is fixed, but not exactly what.
Please be more specific.

https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md#format-git-commit-messages

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/file/FileHiveMetastore.java [hunk 2]
```diff
@@ -736,26 +737,54 @@ public synchronized void commentTable(String databaseName, String tableName, Opt
...
-                if (fieldSchema.getName().equals(columnName)) {
-                    newDataColumns.add(new Column(columnName, fieldSchema.getType(), comment, fieldSchema.getProperties()));
-                }
-                else {
-                    newDataColumns.add(fieldSchema);
-                }
+            Optional<Integer> matchingDataColumn = indexOfColumnWithName(dataColumns, columnName);
+            Optional<Integer> matchingPartitionColumn = indexOfColumnWithName(partitionColumns, columnName);
+
+            if (matchingDataColumn.isPresent() && matchingPartitionColumn.isPresent()) {
```
Comment (findinpath) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/file/FileHiveMetastore.java:747]: when can this happen?

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/file/FileHiveMetastore.java [hunk 3]
```diff
@@ -736,26 +737,54 @@ public synchronized void commentTable(String databaseName, String tableName, Opt
...
+                    .withPartitionColumns(currentVersion, matchingPartitionColumn.map(index -> setColumnCommentForIndex(partitionColumns, index, comment)).orElse(partitionColumns));
         });
     }
 
+    private static Optional<Integer> indexOfColumnWithName(List<Column> columns, String columnName)
+    {
+        Optional<Integer> index = Optional.empty();
+        for (int i = 0; i < columns.size(); i++) {
+            if (columns.get(i).getName().equals(columnName)) {
+                index.ifPresent(ignored -> {
```
Comment (findinpath) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/file/FileHiveMetastore.java:765]: when can this situation occur?
is it worth to handle it?

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/file/FileHiveMetastore.java (hunk 1 above)
Comment (findinpath) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/file/FileHiveMetastore.java:739]: it seems to me that the message should be

"Support altering partition column comments in Hive file metastore"

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java [hunk 4]
```diff
@@ -9010,6 +9010,75 @@ public void testCollidingMixedCaseProperty()
         assertUpdate("DROP TABLE %s".formatted(tableName));
     }
 
+    @Test
+    public void testCommentViaSeparateCommentOnColumnStatement()
+    {
+        String table = "test_predicate_pushdown_" + randomNameSuffix();
```
Comment (ebyhr) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java:9016]: The table name seems wrong. Same for another test method.

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java [hunk 5]
```diff
@@ -9010,6 +9010,75 @@ public void testCollidingMixedCaseProperty()
...
+                   format = 'PARQUET',
+                   partitioned_by = ARRAY['partition_column']
+                )
+                """.formatted(table));
+
+        assertUpdate("COMMENT ON COLUMN %s.regular_column IS 'This is a regular column comment.'".formatted(table));
+        assertUpdate("COMMENT ON COLUMN %s.partition_column IS 'This is a partition column comment.'".formatted(table));
+        assertUpdate("COMMENT ON TABLE %s IS 'This is a table comment.'".formatted(table));
+
+        assertThat((String) getQueryRunner().execute("SHOW CREATE TABLE " + table).getOnlyValue())
```
Comment (ebyhr) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java:9033]: There's no need to validate the entire table definition. I would recommend using `BaseConnectorTest#getTableComment` method instead. 

 Same for another test method.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/file/FileHiveMetastore.java (hunk 2 above)
Comment (ssheikin) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/file/FileHiveMetastore.java:747]: I don't know. I've copied it from Glue meta store. Better safe than sorry.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/file/FileHiveMetastore.java (hunk 3 above)
Comment (ssheikin) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/file/FileHiveMetastore.java:765]: I don't know. I've copied it from Glue meta store. Better safe than sorry.

**Comment by findinpath** [comment]
@ssheikin  pls adjust the release notes in the description to be displayed correctly

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestComments.java [hunk 6]
```diff
@@ -142,10 +142,12 @@ public void testCommentColumn()
                         "CREATE TABLE hive.default.%s (\n" +
                         "   c1 bigint COMMENT 'test comment',\n" +
                         "   c2 bigint COMMENT '',\n" +
-                        "   c3 bigint\n" +
+                        "   c3 bigint,\n" +
+                        "   c4 bigint COMMENT 'test partition comment'\n" +
```
Comment (findinpath) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestComments.java:146]: nit/optional: Let's name `c4` as `part` / `part_col`

Comment (ssheikin) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestComments.java:146]: I prefer to keep it in sync with already existing code.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/BridgingHiveMetastore.java [hunk 7]
```diff
@@ -336,7 +341,14 @@ public void commentColumn(String databaseName, String tableName, String columnNa
         io.trino.hive.thrift.metastore.Table table = delegate.getTable(databaseName, tableName)
                 .orElseThrow(() -> new TableNotFoundException(new SchemaTableName(databaseName, tableName)));
 
-        for (FieldSchema fieldSchema : table.getSd().getCols()) {
+        List<FieldSchema> fieldSchemas = ImmutableList.<FieldSchema>builder()
+                .addAll(table.getPartitionKeys())
+                .addAll(table.getSd().getCols())
+                .build();
+
+        verifyDistinctNames(fieldSchemas);
```
Comment (ssheikin) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/BridgingHiveMetastore.java:349]: @ebyhr does it make sense for Thrift?

**Comment by ssheikin** [comment]
@findinpath @ebyhr @alexjo2144
Comments addressed. Please accept.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/BridgingHiveMetastore.java (hunk 7 above)
Comment (findinpath) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/BridgingHiveMetastore.java:349]: Tried at the HMS database level to perform the change of having two namesake columns of a table without succcess:

```
MariaDB [metastore]> select * from COLUMNS_V2 where CD_ID = 56;
+-------+---------+-------------+-----------+-------------+
| CD_ID | COMMENT | COLUMN_NAME | TYPE_NAME | INTEGER_IDX |
+-------+---------+-------------+-----------+-------------+
|    56 | NULL    | c1          | int       |           0 |
|    56 | NULL    | c2          | int       |           1 |
|    56 | NULL    | c3          | int       |           2 |
+-------+---------+-------------+-----------+-------------+
3 rows in set (0.00 sec)

MariaDB [metastore]> update COLUMNS_V2 set COLUMN_NAME = 'c2' where COLUMN_NAME = 'c1' and CD_ID = '56';
ERROR 1062 (23000): Duplicate entry '56-c2' for key 'PRIMARY'
MariaDB [metastore]> show create table COLUMNS_V2;
+------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table      | Create Table                                                                                                                                                                                                                                                                                                                                                                                                                                                                        |
+------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| COLUMNS_V2 | CREATE TABLE `COLUMNS_V2` (
  `CD_ID` bigint(20) NOT NULL,
  `COMMENT` varchar(256) CHARACTER SET latin1 COLLATE latin1_bin DEFAULT NULL,
  `COLUMN_NAME` varchar(767) CHARACTER SET latin1 COLLATE latin1_bin NOT NULL,
  `TYPE_NAME` mediumtext,
  `INTEGER_IDX` int(11) NOT NULL,
  PRIMARY KEY (`CD_ID`,`COLUMN_NAME`),
  KEY `COLUMNS_V2_N49` (`CD_ID`),
  CONSTRAINT `COLUMNS_V2_FK1` FOREIGN KEY (`CD_ID`) REFERENCES `CDS` (`CD_ID`)
) ENGINE=InnoDB DEFAULT CHARSET=latin1 |

```

I'd say that also for Thrift the check is not necessary.

Comment (ssheikin) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/BridgingHiveMetastore.java:349]: @findinpath thank you for checking that! 
I'm having my hands on removing that check, but I want to check one concern before: 
Do you think we can assume that constraint on column names will be created by all supported thrift metastores trino works with? e.g. different distributions?

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestComments.java [hunk 8]
```diff
@@ -154,7 +156,12 @@ public void testCommentColumn()
...
+                        ")\n" +
+                        "WITH (\n" +
+                        "   format = 'RCBINARY',\n" +
+                        "   partitioned_by = ARRAY['c4']\n" +
```
Comment (ebyhr) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestComments.java:164]: Is the following `\\E(?s:.*)` pattern still required? Same for other places.

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java [hunk 9]
```diff
@@ -9010,6 +9010,56 @@ public void testCollidingMixedCaseProperty()
...
+    {
+        String table = "test_comment_on_" + randomNameSuffix();
+
+        assertUpdate("""
+                CREATE TABLE hive.tpch.%s (
+                   regular_column date,
+                   partition_column date
+                )
+                WITH (
+                   format = 'PARQUET',
```
Comment (ebyhr) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java:9024]: No need to specify format table property. Same for another test method.

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java [hunk 10]
```diff
@@ -9010,6 +9010,56 @@ public void testCollidingMixedCaseProperty()
         assertUpdate("DROP TABLE %s".formatted(tableName));
     }
 
+    @Test
+    public void testCommentViaSeparateCommentOnColumnStatement()
```
Comment (ebyhr) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java:9014]:
```suggestion
    public void testCommentOnPartitionColumn()
```

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java [hunk 11]
```diff
@@ -9010,6 +9010,56 @@ public void testCollidingMixedCaseProperty()
...
+        assertUpdate("COMMENT ON COLUMN %s.partition_column IS 'partition column comment'".formatted(table));
+        assertThat(getColumnComment(table, "partition_column")).isEqualTo("partition column comment");
+        assertUpdate("COMMENT ON TABLE %s IS 'table comment'".formatted(table));
+        assertThat(getTableComment("hive", "tpch", table)).isEqualTo("table comment");
+
+        assertUpdate("DROP TABLE " + table);
+    }
+
+    @Test
+    public void testCommentViaCreateTableStatement()
```
Comment (ebyhr) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java:9040]:
```suggestion
    public void testCreateTableWithPartitionColumnComment()
```

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java [hunk 12]
```diff
@@ -9010,6 +9010,56 @@ public void testCollidingMixedCaseProperty()
...
+    {
+        String table = "test_comment_on_" + randomNameSuffix();
```
Comment (ebyhr) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java:9042]: This test doesn't run `COMMENT ON` statement. I would change to `test_create_table_partition_column_comment_` or something.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/BridgingHiveMetastore.java [hunk 13]
```diff
@@ -336,7 +337,12 @@ public void commentColumn(String databaseName, String tableName, String columnNa
         io.trino.hive.thrift.metastore.Table table = delegate.getTable(databaseName, tableName)
                 .orElseThrow(() -> new TableNotFoundException(new SchemaTableName(databaseName, tableName)));
 
-        for (FieldSchema fieldSchema : table.getSd().getCols()) {
+        List<FieldSchema> fieldSchemas = ImmutableList.<FieldSchema>builder()
+                .addAll(table.getPartitionKeys())
+                .addAll(table.getSd().getCols())
```
Comment (ebyhr) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/BridgingHiveMetastore.java:342]: nit: I would replace the order as we usually process data columns first. 
```suggestion
                .addAll(table.getSd().getCols())
                .addAll(table.getPartitionKeys())
```

**Review by ebyhr — COMMENTED** [review]
Almost good to me. Does Hive support setting column comments on partitioned columns with Thrift metastore? I know we can set comments on data columns using `ALTER TABLE ... CHANGE ...`

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestComments.java (hunk 8 above)
Comment (ssheikin) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestComments.java:164]: This is an existing code. Please let me know if you want me to improve it.

**Comment by ssheikin** [comment]
> Almost good to me. Does Hive support setting column comments on partitioned columns with Thrift metastore? I know we can set comments on data columns using `ALTER TABLE ... CHANGE ...`

No it's not possible to change comment for partition column.
https://stackoverflow.com/a/57871346
However, hive is able to show comment for partition column which is changed by trino.

**Comment by ssheikin** [comment]
@ebyhr all comments are addressed. please review.

**Review by findinpath — APPROVED** [review]

**Comment by ssheikin** [comment]
@ebyhr could you PTAL?

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestComments.java [hunk 14]
```diff
@@ -138,62 +138,39 @@ protected String getColumnComment(String schemaName, String tableName, String co
...
+                   c4 bigint COMMENT 'test partition comment',
+                   c5 bigint COMMENT '',
+                   c6 bigint
+                )
+                WITH (
+                   partitioned_by = ARRAY['c4', 'c5', 'c6']
+                )
+                """.formatted(COMMENT_COLUMN_NAME));
+        assertThat(getColumnComment("default", COMMENT_COLUMN_NAME, "c1")).isEqualTo("test comment");
+        assertThat(getColumnComment("default", COMMENT_COLUMN_NAME, "c2")).isEqualTo("");
```
Comment (ebyhr) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestComments.java:155]:
```suggestion
        assertThat(getColumnComment("default", COMMENT_COLUMN_NAME, "c2")).isEmpty();
```

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestComments.java [hunk 15]
```diff
@@ -138,62 +138,39 @@ protected String getColumnComment(String schemaName, String tableName, String co
...
+        assertThat(getColumnComment("default", COMMENT_COLUMN_NAME, "c3")).isNull();
+        assertThat(getColumnComment("default", COMMENT_COLUMN_NAME, "c4")).isEqualTo("test partition comment");
+        assertThat(getColumnComment("default", COMMENT_COLUMN_NAME, "c5")).isEqualTo("");
```
Comment (ebyhr) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestComments.java:158]:
```suggestion
        assertThat(getColumnComment("default", COMMENT_COLUMN_NAME, "c5")).isEmpty();
```

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java [hunk 16]
```diff
@@ -9009,6 +9009,50 @@ public void testCollidingMixedCaseProperty()
         assertUpdate("DROP TABLE %s".formatted(tableName));
     }
 
+    @Test
+    public void testTableComments()
+    {
+        String table = "test_create_table_with_comments_" + randomNameSuffix();
```
Comment (ebyhr) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java:9015]: I feel the test name & the variable is different from the actual assertions.

**Review by ebyhr — APPROVED** [review]

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/file/FileHiveMetastore.java [hunk 17]
```diff
@@ -734,24 +734,18 @@ public synchronized void commentTable(String databaseName, String tableName, Opt
...
-                }
-            }
+        alterTable(databaseName, tableName, table -> table
+                .withDataColumns(currentVersion, updateColumnComment(table.getDataColumns(), columnName, comment))
+                .withPartitionColumns(currentVersion, updateColumnComment(table.getPartitionColumns(), columnName, comment)));
+    }
 
-            return oldTable.withDataColumns(currentVersion, newDataColumns.build());
-        });
+    private static List<Column> updateColumnComment(List<Column> originalColumns, String columnName, Optional<String> comment)
```
Comment (wendigo) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/file/FileHiveMetastore.java:742]: I don't like this implementation

Comment (wendigo) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/file/FileHiveMetastore.java:742]:
```s
        return Lists.transform(originalColumns, column -> column.getName().equals(columnName)
                ? new Column(column.getName(), column.getType(), comment, column.getProperties())
                : column);
```

**Review by wendigo — APPROVED** [review]

**Pull request merged**